docs/design/loadpath.md

```diff
@@ -0,0 +1,284 @@
+the signature when the program is loaded. LSM hooks can the be used to further
+apply policy to the program that's being loaded.
+
+This prevents a bad actor from modifying the eBPF bytecode in a program, since
+the signature will no longer be valid.
+
+This proposal is still a work in progress, but it's a promising development.
+In particular, the integration with the IMA subsystem is very interesting as it
+opens up the possibility of remote attestations using Keylime. Not only that,
+the more I learn about IMAs capabilities, the more I think it's a good fit for
+eBPF.
+
+There are however some drawbacks to this approach:
+
+- Fsverity is limited to ext4, btrfs and f2fs filesystems.
+- Support for fsverity on overlayfs is enabled via mount option - additional work
+  may be required to support this in container runtimes.
+- This approach is less effective on static binaries, since if they were to,
+  for example, link to libbpf, then you would also need libbpf to be signed
+  with fsverity to avoid any potential tampering. This could be mitigated by
+  having relocations performed by the kernel instead.
+- Some investigation is needed to see whether the fsverity xattrs are preserved
+  when a file is copied into a container image, or signed there as part of the
+  build process.
+
+[BPF Signing using fsverity and LSM Gatekeepers]: http://vger.kernel.org/bpfconf2023_material/Lorenz_Bauer_-_BPF_signing_using_fsverity_and_LSM_gatekeeper.pdf
+
+#### BPF Signatures using eBPF Skeletons
+
+[BPF Signatures] proposal is being championed by KP Singh.
+This approach involves generating an eBPF "light skeleton" that contains:
+
+- The eBPF bytecode
+- A recording of the syscalls that would be used to load the program
+
+This light skeleton would then be signed, and the signature would be verified
+by the kernel when the program is loaded. eBPF LSM or IMA could be used to
+further apply policy to the program that's being loaded.
+
+This uses the eBPF program type "BPF_PROG_TYPE_SYSCALL", which effectively uses
+eBPF to replay the syscalls and load the program.
+
+There are some drawbacks to this approach also:
+
+- CO:RE relocations occur when you generate the skeleton. As such, you would
+  need to generate a skeleton for each kernel version that you want to support
+  which partially defeats the point of CO:RE in the first place.
+- Light Skeletons are not supported for all program types
+- I don't believe that this approach would work with BPF Tokens since the
+  token cannot be embedded in the skeleton.
+
+[BPF Signatures]: https://lpc.events/event/16/contributions/1357/attachments/1045/1999/BPF%20Signatures.pdf
+
+#### Signing Conclusion
+
+Both of these approaches have merits and some drawbacks.
+In an ideal world, we would have both of these approaches available to us in the
+following way:
+
+- All CO:RE relocations occur in the kernel - making libbpf and other libraries
+  that use CO:RE much simpler.
```

Comment:
...by moving that complexity into the kernel so it runs with ring 0 privileges? Not sure that's an improvement...